**复习：**在前面我们已经学习了Pandas基础，知道利用Pandas读取csv数据的增删查改，今天我们要学习的就是**探索性数据分析**，主要介绍如何利用Pandas进行排序、算术计算以及计算描述函数describe()的使用。

# 1 第一章：探索性数据分析

#### 开始之前，导入numpy、pandas包和数据

In [1]:
#加载所需的库
import numpy as np
import pandas as pd

In [2]:
#载入之前保存的train_chinese.csv数据，关于泰坦尼克号的任务，我们就使用这个数据
text = pd.read_csv('train_chinese.csv')
text.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

### 1.6 了解你的数据吗？
教材《Python for Data Analysis》第五章

#### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [3]:
# 具体请看《利用Python进行数据分析》第五章 排序和排名 部分

#自己构建一个都为数字的DataFrame数据
frame = pd.DataFrame(np.arange(8).reshape((2, 4)), 
                     index=['2', '1'], 
                     columns=['d', 'a', 'b', 'c'])
frame


d  a  b  c
2  0  1  2  3
1  4  5  6  7

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象 

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行


In [4]:
# 大多数时候我们都是想根据列的值来排序,所以，将你构建的DataFrame中的数据根据某一列，升序排列
frame.sort_values(by='c', ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

可以看到sort_values这个函数中by参数指向要排列的列，ascending参数指向排序的方式（升序还是降序）

【总结】下面将不同的排序方式做一个小总结

In [5]:
# 让行索引升序排序
frame.sort_index()

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [6]:
# 让列索引升序排序
frame.sort_index(axis=1)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

In [7]:
# 让列索引降序排序
frame.sort_index(axis=1, ascending=False)

d  c  b  a
2  0  3  2  1
1  4  7  6  5

In [8]:
# 让任选两列数据同时降序排序
frame.sort_values(by=['a', 'c'])

d  a  b  c
2  0  1  2  3
1  4  5  6  7

#### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从数据中你能发现什么

In [9]:
text.sort_values(by=['票价', '年龄'], ascending=False).head(3)

乘客ID  是否幸存  仓位等级                                  姓名      性别    年龄  \
679   680     1     1  Cardeza, Mr. Thomas Drake Martinez    male  36.0   
258   259     1     1                    Ward, Miss. Anna  female  35.0   
737   738     1     1              Lesurer, Mr. Gustave J    male  35.0   

     兄弟姐妹个数  父母子女个数      船票信息        票价           客舱 登船港口  
679       0       1  PC 17755  512.3292  B51 B53 B55    C  
258       0       0  PC 17755  512.3292          NaN    C  
737       0       0  PC 17755  512.3292         B101    C

【思考】如果继续研究相关性的话可以画一个系数矩阵图进行研究


#### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [10]:
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [11]:
frame1_b

a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

In [12]:
#将frame_a和frame_b进行相加
frame1_a + frame1_b

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。<br>
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

#### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [13]:
max(text['兄弟姐妹个数'] + text['父母子女个数'])

10

是的，如上，很简单，我们只需找出兄弟姐妹个数和父母子女个数之和最大的数就行，先让这两列相加返回一个DataFrame，然后用max函数求出最大值，当然你还可以想出很多方法和思考角度，欢迎你来说出你的看法。

#### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [14]:
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [15]:
# 调用 describe 函数，观察frame2的数据基本信息

frame2.describe()

'''
count : 样本数据大小
mean : 样本数据的平均值
std : 样本数据的标准差
min : 样本数据的最小值
25% : 样本数据25%的时候的值
50% : 样本数据50%的时候的值
75% : 样本数据75%的时候的值
max : 样本数据的最大值
'''

'\ncount : 样本数据大小\nmean : 样本数据的平均值\nstd : 样本数据的标准差\nmin : 样本数据的最小值\n25% : 样本数据25%的时候的值\n50% : 样本数据50%的时候的值\n75% : 样本数据75%的时候的值\nmax : 样本数据的最大值\n'

#### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [16]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''
text['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

【思考】从上面数据我们可以看出，
一共有891个票价数据，
平均值约为：32.20，
标准差约为49.69，说明票价波动特别大，
25%的人的票价是低于7.91的，50%的人的票价低于14.45，75%的人的票价低于31.00，
票价最大值约为512.33，最小值为0。
<br>当然，这只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

In [17]:
'''
通过上面的例子，我们再看看泰坦尼克号数据集中 父母子女个数 这列数据的基本统计数据，然后可以说出你的想法
'''
text['父母子女个数'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: 父母子女个数, dtype: float64